File: aic_data_collection/aic_data_collection/auto_data_collector.py
"""Main orchestrator node for automated data collection."""

import threading
import time
from typing import Optional

import rclpy
from rclpy.action import ActionClient
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from rclpy.callback_groups import ReentrantCallbackGroup

from aic_task_interfaces.action import InsertCable
from aic_task_interfaces.msg import Task
from lifecycle_msgs.srv import ChangeState
from lifecycle_msgs.msg import Transition
from sensor_msgs.msg import Image
from std_msgs.msg import String
import tf2_ros

from .completion_monitor import CompletionResult, check_tf_completion
from .homing_manager import HomingManager
from .rosbag_manager import RosbagManager
from .scene_manager import SceneManager
from .trial_config_provider import StaticTrialProvider


class AutoDataCollector(Node):
    """Orchestrates automated data collection across multiple episodes."""

    def __init__(self):
        super().__init__("auto_data_collector")

        # Declare parameters
        self.declare_parameter("config_file", "")
        self.declare_parameter("trial_mode", "static")
        self.declare_parameter("trials", [""])
        self.declare_parameter("target_episodes", 10)
        self.declare_parameter("max_attempts", 0)  # 0 = target × 3
        self.declare_parameter("task_timeout_sec", 180.0)
        self.declare_parameter("bag_output_dir", "bags")
        self.declare_parameter("completion_distance_threshold", 0.005)
        self.declare_parameter("completion_orientation_threshold", 0.1)

        # Read parameters
        config_file = self.get_parameter("config_file").get_parameter_value().string_value
        if not config_file:
            self.get_logger().fatal("config_file parameter is required")
            raise ValueError("config_file parameter is required")

        trial_names_param = self.get_parameter("trials").get_parameter_value().string_array_value
        trial_names = [t for t in trial_names_param if t] or None

        self._target_episodes = self.get_parameter("target_episodes").get_parameter_value().integer_value
        max_attempts = self.get_parameter("max_attempts").get_parameter_value().integer_value
        self._max_attempts = max_attempts if max_attempts > 0 else self._target_episodes * 3
        self._task_timeout = self.get_parameter("task_timeout_sec").get_parameter_value().double_value
        bag_dir = self.get_parameter("bag_output_dir").get_parameter_value().string_value
        self._dist_threshold = self.get_parameter("completion_distance_threshold").get_parameter_value().double_value
        self._orient_threshold = self.get_parameter("completion_orientation_threshold").get_parameter_value().double_value

        # Initialize components
        trial_mode = self.get_parameter("trial_mode").get_parameter_value().string_value
        if trial_mode == "dynamic":
            from .trial_config_provider import DynamicTrialProvider
            base_trial = trial_names[0] if trial_names else None
            if not base_trial:
                raise ValueError("dynamic mode requires at least one trial name as base_trial")
            self._trial_provider = DynamicTrialProvider(config_file, base_trial=base_trial)
        else:
            self._trial_provider = StaticTrialProvider(config_file, trials=trial_names)

        self._rosbag = RosbagManager(output_dir=bag_dir)
        self._scene = SceneManager(self)
        self._homing = HomingManager(self, self._trial_provider.home_joint_positions)

        # Camera dummy subscribers (activates lazy bridge for video topics)
        self._camera_subs = []
        for topic in ["/left_camera/image", "/center_camera/image", "/right_camera/image"]:
            sub = self.create_subscription(Image, topic, lambda msg: None, 1)
            self._camera_subs.append(sub)
        self.get_logger().info("Camera dummy subscribers created (lazy bridge activated)")

        # InsertCable action client
        self._cb_group = ReentrantCallbackGroup()
        self._insert_cable_client = ActionClient(
            self, InsertCable, "/insert_cable", callback_group=self._cb_group
        )

        # Completion monitoring state
        self._insertion_event_received = False
        self._insertion_event_sub = self.create_subscription(
            String, "/scoring/insertion_event", self._on_insertion_event, 10
        )

        # TF for completion check
        self._tf_buffer = tf2_ros.Buffer()
        self._tf_listener = tf2_ros.TransformListener(self._tf_buffer, self)

        # aic_model lifecycle client
        self._model_change_state = self.create_client(ChangeState, "/aic_model/change_state")

        # Stats
        self._successful = 0
        self._failed = 0
        self._fail_reasons: dict[str, int] = {}

    def _on_insertion_event(self, msg: String) -> None:
        self._insertion_event_received = True
        self.get_logger().info(f"Insertion event received: {msg.data}")

    def _activate_model(self) -> bool:
        """Configure and activate aic_model lifecycle node."""
        if not self._model_change_state.wait_for_service(timeout_sec=30.0):
            self.get_logger().error("aic_model lifecycle service not available")
            return False

        # Configure
        req = ChangeState.Request()
        req.transition = Transition(id=Transition.TRANSITION_CONFIGURE)
        future = self._model_change_state.call_async(req)
        self._poll_future(future, 30.0)
        if not future.done() or not future.result().success:
            self.get_logger().error("Failed to configure aic_model")
            return False

        # Activate
        req = ChangeState.Request()
        req.transition = Transition(id=Transition.TRANSITION_ACTIVATE)
        future = self._model_change_state.call_async(req)
        self._poll_future(future, 30.0)
        if not future.done() or not future.result().success:
            self.get_logger().error("Failed to activate aic_model")
            return False

        self.get_logger().info("aic_model activated")
        return True

    def _send_insert_cable(self, task_config: dict) -> Optional[bool]:
        """Send InsertCable action goal. Returns True=success, False=fail, None=timeout."""
        if not self._insert_cable_client.wait_for_action_server(timeout_sec=10.0):
            self.get_logger().error("InsertCable action server not available")
            return False

        goal = InsertCable.Goal()
        goal.task = Task()
        goal.task.id = task_config.get("cable_name", "task_1")
        goal.task.cable_type = task_config["cable_type"]
        goal.task.cable_name = task_config["cable_name"]
        goal.task.plug_type = task_config["plug_type"]
        goal.task.plug_name = task_config["plug_name"]
        goal.task.port_type = task_config["port_type"]
        goal.task.port_name = task_config["port_name"]
        goal.task.target_module_name = task_config["target_module_name"]
        goal.task.time_limit = int(self._task_timeout)

        send_goal_future = self._insert_cable_client.send_goal_async(goal)
        self._poll_future(send_goal_future, 10.0)
        if not send_goal_future.done():
            return None
        goal_handle = send_goal_future.result()
        if not goal_handle or not goal_handle.accepted:
            self.get_logger().error("InsertCable goal rejected")
            return False

        result_future = goal_handle.get_result_async()
        self._poll_future(result_future, self._task_timeout)
        if not result_future.done():
            self.get_logger().warn("InsertCable action timed out, cancelling")
            goal_handle.cancel_goal_async()
            return None
        result = result_future.result()
        return result.result.success if result else False

    def _check_completion(self, task_config: dict) -> CompletionResult:
        """Evaluate task completion using insertion_event + TF."""
        if not self._insertion_event_received:
            return CompletionResult.TIMEOUT

        # TF check
        plug_frame = f"{task_config['cable_name']}/{task_config['plug_name']}_link"
        port_frame = f"task_board/{task_config['target_module_name']}/{task_config['port_name']}_link"
        try:
            plug_tf = self._tf_buffer.lookup_transform("base_link", plug_frame, tf2_ros.Time())
            port_tf = self._tf_buffer.lookup_transform("base_link", port_frame, tf2_ros.Time())
            plug_pos = (plug_tf.transform.translation.x, plug_tf.transform.translation.y, plug_tf.transform.translation.z)
            port_pos = (port_tf.transform.translation.x, port_tf.transform.translation.y, port_tf.transform.translation.z)
            plug_quat = (plug_tf.transform.rotation.x, plug_tf.transform.rotation.y, plug_tf.transform.rotation.z, plug_tf.transform.rotation.w)
            port_quat = (port_tf.transform.rotation.x, port_tf.transform.rotation.y, port_tf.transform.rotation.z, port_tf.transform.rotation.w)

            if check_tf_completion(plug_pos, port_pos, plug_quat, port_quat,
                                   self._dist_threshold, self._orient_threshold):
                return CompletionResult.SUCCESS
            else:
                self.get_logger().warn("insertion_event received but TF check failed")
                return CompletionResult.PARTIAL
        except Exception as e:
            self.get_logger().warn(f"TF lookup failed: {e}, treating as PARTIAL")
            return CompletionResult.PARTIAL

    def _poll_future(self, future, timeout: float) -> None:
        """Poll a future until done or timeout (avoids executor deadlock)."""
        start = time.monotonic()
        while not future.done() and (time.monotonic() - start) < timeout:
            time.sleep(0.05)

    def run_collection(self) -> None:
        """Main collection loop."""
        self.get_logger().info(f"Starting collection: target={self._target_episodes}, max_attempts={self._max_attempts}")

        # Activate aic_model once
        if not self._activate_model():
            self.get_logger().fatal("Cannot activate aic_model, aborting")
            return

        attempt = 0
        while self._successful < self._target_episodes and attempt < self._max_attempts:
            attempt += 1
            trial = self._trial_provider.get_next_trial()
            trial_id = trial["trial_id"]
            episode_id = f"{trial_id}_ep{attempt:04d}"

            self.get_logger().info(f"\n{'='*50}")
            self.get_logger().info(f"Episode {attempt}/{self._max_attempts} ({self._successful}/{self._target_episodes} success) - {trial_id}")
            self.get_logger().info(f"{'='*50}")

            # Reset completion state
            self._insertion_event_received = False

            # 1. Spawn scene
            if not self._scene.spawn_scene(trial):
                self._record_failure("spawn_failed")
                continue

            # 2. Start recording
            self._rosbag.start(episode_id)

            # 3. Execute task (first task in trial)
            first_task_key = list(trial["tasks"].keys())[0]
            task_config = trial["tasks"][first_task_key]
            action_result = self._send_insert_cable(task_config)

            # 4. Check completion
            if action_result is None:
                completion = CompletionResult.TIMEOUT
            elif action_result is False:
                completion = CompletionResult.ACTION_FAILED
            else:
                completion = self._check_completion(task_config)

            # 5. Stop recording
            success = completion.is_success()
            bag_path = self._rosbag.stop(success=success)
            self.get_logger().info(f"Bag saved: {bag_path} (result: {completion.value})")

            # 6. Update stats
            if success:
                self._successful += 1
            else:
                self._record_failure(completion.value)

            # 7. Despawn + home
            self._scene.despawn_scene()
            if not self._homing.home_robot():
                self.get_logger().error("Homing failed!")

        self._print_summary()

    def _record_failure(self, reason: str) -> None:
        self._failed += 1
        self._fail_reasons[reason] = self._fail_reasons.get(reason, 0) + 1

    def _print_summary(self) -> None:
        total = self._successful + self._failed
        self.get_logger().info(f"\n{'='*50}")
        self.get_logger().info("=== Data Collection Summary ===")
        self.get_logger().info(f"Total attempts: {total}")
        target_msg = "target reached" if self._successful >= self._target_episodes else "target NOT reached"
        self.get_logger().info(f"Successful episodes: {self._successful} / {self._target_episodes} ({target_msg})")
        self.get_logger().info(f"Failed episodes: {self._failed}")
        for reason, count in self._fail_reasons.items():
            self.get_logger().info(f"  - {reason}: {count}")


def main(args=None):
    rclpy.init(args=args)
    node = AutoDataCollector()
    executor = MultiThreadedExecutor()
    executor.add_node(node)
    spin_thread = threading.Thread(target=executor.spin, daemon=True)
    spin_thread.start()

    try:
        node.run_collection()
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: aic_engine/src/auto_data_collector.cpp
#include "auto_data_collector.hpp"

#include <chrono>
#include <iomanip>
#include <sstream>

#include "aic_task_interfaces/msg/task.hpp"
#include "lifecycle_msgs/msg/state.hpp"
#include "lifecycle_msgs/msg/transition.hpp"

namespace aic {

AutoDataCollector::AutoDataCollector()
    : rclcpp::Node("auto_data_collector") {
  // Declare parameters
  declare_parameter("config_file", std::string(""));
  declare_parameter("trial_mode", std::string("static"));
  declare_parameter("trials", std::vector<std::string>{""});
  declare_parameter("target_episodes", 10);
  declare_parameter("max_attempts", 0);
  declare_parameter("task_timeout_sec", 180.0);
  declare_parameter("bag_output_dir", std::string("bags"));
  declare_parameter("completion_distance_threshold", 0.005);
  declare_parameter("completion_orientation_threshold", 0.1);

  // Read parameters
  std::string config_file = get_parameter("config_file").as_string();
  if (config_file.empty()) {
    // Try alternate parameter name used by aic_engine
    declare_parameter("config_file_path", std::string(""));
    config_file = get_parameter("config_file_path").as_string();
  }
  if (config_file.empty()) {
    RCLCPP_FATAL(get_logger(), "config_file parameter is required");
    throw std::runtime_error("config_file parameter is required");
  }

  auto trial_names_param = get_parameter("trials").as_string_array();
  std::vector<std::string> trial_names;
  for (const auto& t : trial_names_param) {
    if (!t.empty()) trial_names.push_back(t);
  }

  target_episodes_ = get_parameter("target_episodes").as_int();
  int max_att = get_parameter("max_attempts").as_int();
  max_attempts_ = max_att > 0 ? max_att : target_episodes_ * 3;
  task_timeout_sec_ = get_parameter("task_timeout_sec").as_double();
  std::string bag_dir = get_parameter("bag_output_dir").as_string();
  dist_threshold_ = get_parameter("completion_distance_threshold").as_double();
  orient_threshold_ = get_parameter("completion_orientation_threshold").as_double();

  // Load root config for spawn_scene
  root_config_ = YAML::LoadFile(config_file);

  // Initialize trial provider
  std::string trial_mode = get_parameter("trial_mode").as_string();
  if (trial_mode == "dynamic") {
    trial_provider_ = std::make_unique<DynamicTrialProvider>(config_file, trial_names);
  } else {
    trial_provider_ = std::make_unique<StaticTrialProvider>(config_file, trial_names);
  }

  // Initialize components
  tf_helper_ = std::make_unique<TfHelper>(this);
  scene_spawner_ = std::make_unique<SceneSpawner>(this, *tf_helper_);
  homing_helper_ = std::make_unique<HomingHelper>(this, trial_provider_->home_joint_positions());
  rosbag_manager_ = std::make_unique<RosbagManager>(bag_dir);

  // Camera dummy subscribers (activates lazy bridge)
  for (const auto& topic : {"/left_camera/image", "/center_camera/image", "/right_camera/image"}) {
    auto sub = create_subscription<sensor_msgs::msg::Image>(
        topic, 1, [](const sensor_msgs::msg::Image::SharedPtr) {});
    camera_subs_.push_back(sub);
  }
  RCLCPP_INFO(get_logger(), "Camera dummy subscribers created (lazy bridge activated)");

  // InsertCable action client
  action_client_ = rclcpp_action::create_client<InsertCableAction>(this, "/insert_cable");

  // Insertion event subscription
  insertion_event_sub_ = create_subscription<std_msgs::msg::String>(
      "/scoring/insertion_event", 10,
      [this](const std_msgs::msg::String::SharedPtr msg) {
        insertion_event_received_ = true;
        RCLCPP_INFO(get_logger(), "Insertion event received: %s", msg->data.c_str());
      });

  // aic_model lifecycle clients
  model_get_state_ = create_client<lifecycle_msgs::srv::GetState>("/aic_model/get_state");
  model_change_state_ = create_client<lifecycle_msgs::srv::ChangeState>("/aic_model/change_state");
}

int AutoDataCollector::get_model_state() {
  if (!model_get_state_->wait_for_service(std::chrono::seconds(10))) {
    return -1;
  }
  auto req = std::make_shared<lifecycle_msgs::srv::GetState::Request>();
  auto future = model_get_state_->async_send_request(req);
  if (!poll_future(future, 10.0)) return -1;
  auto response = future.get();
  if (!response) return -1;
  return response->current_state.id;
}

bool AutoDataCollector::activate_model() {
  if (!model_change_state_->wait_for_service(std::chrono::seconds(30))) {
    RCLCPP_ERROR(get_logger(), "aic_model lifecycle service not available");
    return false;
  }

  int current_state = get_model_state();
  if (current_state == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
    RCLCPP_INFO(get_logger(), "aic_model already active");
    return true;
  }

  if (current_state == lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED ||
      current_state == lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN) {
    auto req = std::make_shared<lifecycle_msgs::srv::ChangeState::Request>();
    req->transition.id = lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE;
    auto future = model_change_state_->async_send_request(req);
    if (!poll_future(future, 30.0) || !future.get()->success) {
      RCLCPP_ERROR(get_logger(), "Failed to configure aic_model");
      return false;
    }
  }

  current_state = get_model_state();
  if (current_state == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE) {
    auto req = std::make_shared<lifecycle_msgs::srv::ChangeState::Request>();
    req->transition.id = lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE;
    auto future = model_change_state_->async_send_request(req);
    if (!poll_future(future, 30.0) || !future.get()->success) {
      RCLCPP_ERROR(get_logger(), "Failed to activate aic_model");
      return false;
    }
  }

  RCLCPP_INFO(get_logger(), "aic_model activated");
  return true;
}

std::optional<bool> AutoDataCollector::send_insert_cable(const YAML::Node& task_config) {
  if (!action_client_->wait_for_action_server(std::chrono::seconds(10))) {
    RCLCPP_ERROR(get_logger(), "InsertCable action server not available");
    return false;
  }

  auto goal = InsertCableAction::Goal();
  goal.task.id = task_config["cable_name"].as<std::string>("task_1");
  goal.task.cable_type = task_config["cable_type"].as<std::string>();
  goal.task.cable_name = task_config["cable_name"].as<std::string>();
  goal.task.plug_type = task_config["plug_type"].as<std::string>();
  goal.task.plug_name = task_config["plug_name"].as<std::string>();
  goal.task.port_type = task_config["port_type"].as<std::string>();
  goal.task.port_name = task_config["port_name"].as<std::string>();
  goal.task.target_module_name = task_config["target_module_name"].as<std::string>();
  goal.task.time_limit = static_cast<uint64_t>(task_timeout_sec_);

  auto send_goal_future = action_client_->async_send_goal(goal);
  if (!poll_future(send_goal_future, 10.0)) return std::nullopt;
  auto goal_handle = send_goal_future.get();
  if (!goal_handle) {
    RCLCPP_ERROR(get_logger(), "InsertCable goal rejected");
    return false;
  }

  auto result_future = action_client_->async_get_result(goal_handle);
  if (!poll_future(result_future, task_timeout_sec_)) {
    RCLCPP_WARN(get_logger(), "InsertCable action timed out, cancelling");
    action_client_->async_cancel_goal(goal_handle);
    return std::nullopt;
  }
  auto result = result_future.get();
  return result.result ? std::optional<bool>(result.result->success) : std::optional<bool>(false);
}

CompletionResult AutoDataCollector::check_completion(const YAML::Node& task_config) {
  if (!insertion_event_received_) {
    return CompletionResult::TIMEOUT;
  }

  std::string plug_frame = task_config["cable_name"].as<std::string>() + "/" +
                           task_config["plug_name"].as<std::string>() + "_link";
  std::string port_frame = "task_board/" +
                           task_config["target_module_name"].as<std::string>() + "/" +
                           task_config["port_name"].as<std::string>() + "_link";

  auto plug_tf = tf_helper_->lookup("base_link", plug_frame);
  auto port_tf = tf_helper_->lookup("base_link", port_frame);

  if (!plug_tf || !port_tf) {
    RCLCPP_WARN(get_logger(), "TF lookup failed, treating as PARTIAL");
    return CompletionResult::PARTIAL;
  }

  Eigen::Vector3d plug_pos(plug_tf->transform.translation.x,
                           plug_tf->transform.translation.y,
                           plug_tf->transform.translation.z);
  Eigen::Vector3d port_pos(port_tf->transform.translation.x,
                           port_tf->transform.translation.y,
                           port_tf->transform.translation.z);
  Eigen::Quaterniond plug_quat(plug_tf->transform.rotation.w,
                               plug_tf->transform.rotation.x,
                               plug_tf->transform.rotation.y,
                               plug_tf->transform.rotation.z);
  Eigen::Quaterniond port_quat(port_tf->transform.rotation.w,
                               port_tf->transform.rotation.x,
                               port_tf->transform.rotation.y,
                               port_tf->transform.rotation.z);

  if (check_tf_completion(plug_pos, port_pos, plug_quat, port_quat,
                          dist_threshold_, orient_threshold_)) {
    return CompletionResult::SUCCESS;
  }
  RCLCPP_WARN(get_logger(), "insertion_event received but TF check failed");
  return CompletionResult::PARTIAL;
}

void AutoDataCollector::record_failure(const std::string& reason) {
  failed_++;
  fail_reasons_[reason]++;
}

void AutoDataCollector::print_summary() {
  int total = successful_ + failed_;
  RCLCPP_INFO(get_logger(), "\n==================================================");
  RCLCPP_INFO(get_logger(), "=== Data Collection Summary ===");
  RCLCPP_INFO(get_logger(), "Total attempts: %d", total);
  std::string target_msg = successful_ >= target_episodes_ ? "target reached" : "target NOT reached";
  RCLCPP_INFO(get_logger(), "Successful episodes: %d / %d (%s)",
              successful_, target_episodes_, target_msg.c_str());
  RCLCPP_INFO(get_logger(), "Failed episodes: %d", failed_);
  for (const auto& [reason, count] : fail_reasons_) {
    RCLCPP_INFO(get_logger(), "  - %s: %d", reason.c_str(), count);
  }
}

void AutoDataCollector::run_collection() {
  RCLCPP_INFO(get_logger(), "Starting collection: target=%d, max_attempts=%d",
              target_episodes_, max_attempts_);

  if (!activate_model()) {
    RCLCPP_FATAL(get_logger(), "Cannot activate aic_model, aborting");
    return;
  }

  int attempt = 0;
  while (successful_ < target_episodes_ && attempt < max_attempts_) {
    attempt++;
    auto trial = trial_provider_->get_next_trial();
    std::string trial_id = trial["trial_id"].as<std::string>();

    auto now = std::chrono::system_clock::now();
    auto time_t_val = std::chrono::system_clock::to_time_t(now);
    struct tm tm_buf;
    localtime_r(&time_t_val, &tm_buf);
    std::ostringstream ts;
    ts << std::put_time(&tm_buf, "%Y%m%d_%H%M%S");
    std::string episode_id = trial_id + "_" + ts.str();

    RCLCPP_INFO(get_logger(), "\n==================================================");
    RCLCPP_INFO(get_logger(), "Episode %d/%d (%d/%d success) - %s",
                attempt, max_attempts_, successful_, target_episodes_, trial_id.c_str());

    // Reset completion state
    insertion_event_received_ = false;

    // 1. Spawn scene
    if (!scene_spawner_->spawn_scene(trial, root_config_)) {
      record_failure("spawn_failed");
      scene_spawner_->despawn_scene();
      homing_helper_->home_robot();
      continue;
    }

    // 2. Start recording
    rosbag_manager_->start_recording(episode_id);

    // 3. Execute task
    auto tasks = trial["tasks"];
    auto first_task_it = tasks.begin();
    auto task_config = first_task_it->second;
    auto action_result = send_insert_cable(task_config);

    // 4. Check completion
    CompletionResult completion;
    if (!action_result.has_value()) {
      completion = CompletionResult::TIMEOUT;
    } else if (!action_result.value()) {
      completion = CompletionResult::ACTION_FAILED;
    } else {
      completion = check_completion(task_config);
    }

    // 5. Stop recording
    bool success = is_success(completion);
    std::string bag_path = rosbag_manager_->stop_recording(success);
    RCLCPP_INFO(get_logger(), "Bag saved: %s (result: %d)", bag_path.c_str(),
                static_cast<int>(completion));

    // 6. Save metadata
    rosbag_manager_->save_metadata(episode_id, trial, success, bag_path);

    // Track episode
    episode_records_.push_back({episode_id, trial_id, success, bag_path});

    // 7. Update stats
    if (success) {
      successful_++;
    } else {
      std::string reason;
      switch (completion) {
        case CompletionResult::TIMEOUT: reason = "timeout"; break;
        case CompletionResult::ACTION_FAILED: reason = "action_failed"; break;
        default: reason = "unknown"; break;
      }
      record_failure(reason);
    }

    // 8. Despawn + home
    scene_spawner_->despawn_scene();
    if (!homing_helper_->home_robot()) {
      RCLCPP_ERROR(get_logger(), "Homing failed!");
    }
  }

  // Save manifest
  if (!episode_records_.empty()) {
    auto manifest_path = rosbag_manager_->save_manifest(episode_records_);
    RCLCPP_INFO(get_logger(), "Manifest saved: %s", manifest_path.c_str());
  }

  print_summary();
}

}  // namespace aic
